src/app/blog/frost-threshold-signatures/page.mdx

```diff
@@ -0,0 +1,346 @@
+
+In most cases, there is a constant conflict between the need to keep the keys safe and the need to constantly access the private key for signing messages.
+
+## Existing solutions
+
+### Local file system
+
+The default way to store a private key is to just store it in a hidden directory in your local file system. While this is not extremely secure, it is still highly preferable to not using encryption at all. In many scenarios, e.g. device loss or theft, this is perfectly fine for low to medium value keypairs.
+
+### Secure key storage
+
+Most modern hardware supports secure storage for private keys. However, access to such secure storage locations is highly platform dependent. It also only works for a limited set of cryptographic primitives, which might not include EdDSA. More fundamentally, while secure storage makes the key relatively inaccessible, it does not protect against key loss. It also does not provide a mechanism for revocation.
+
+### Delegation schemes
+
+With the tools of public key cryptography you can come up with delegation schemes where a rarely used master key is used to delegate to a more frequently used keypair that can be revoked using the master key. This is a very complex topic that would require its own series of blog posts.
+
+## Threshold signatures
+
+I was vaguely aware that something like threshold signatures exist. This is - very roughly speaking - a scheme where you split the private key into multiple parts called *shares*, and need a certain number of these shares to sign a message. Since the shares never have to be in one place, this provides safety in case a single share gets compromised or lost.
+
+What I did not know however is that there exist threshold signature schemes such as FROST that work with Ed25519 such that generated signatures are *fully compatible* with normal Ed25519 signatures. So you can sign a message with a threshold signature scheme and then validate the signature as usual using the ed public key.
+
+This means that such threshold signatures are compatible with existing infrastructure such as [bep-0044] in the mainline DHT, [pkarr], and [nostr].
+
+They are also compatible with all the other places in iroh where we are using ed keypairs.
+
+## Creating key shares
+
+The reason I got interested in threshold signatures is the backwards compatibility.
+
+There are various ways to create key shares. One I find interesting in particular is the ability to just take an *existing* ed private key and generate key shares from it. This way the scheme is not just compatible with Ed25519 signatures in general, but even with existing keypairs. So for example you can use this algorithm for an existing node id, document, or [nostr] account.
+
+Other ways include a scheme that requires a central *trusted dealer*, so basically a place that is considered secure at generation time and a way to securely transfer the shares.
```

Comment:
The first way you define *is* a trusted dealer.
All that the `frost_ed25519::keys::generate_with_dealer` function does is [generate a secret & call `split`](https://docs.rs/frost-core/1.0.0/src/frost_core/keys.rs.html#529-541).

I'd personally consider these two things the same, and call the node the trusted dealer in both cases.

Reply:
Changed the wording. I guess there is some difference because the key never makes it to persistent storage, but the difference is not that big.